### PR TITLE
Crate review

### PR DESCRIPTION
**Pull request by iquerejeta**
Four main changes to the useful version: 
* Updated dependencies
* Used criterion for benches (and moved them to their own file)
* Included documentation
* Changed signature structure. The signature, contrarily to the secret key, does not need to allow for mapping the data to erasable buffer. Therefore, we changed the structure to allow for a more readable description (with explicit elements such as `sigma`, `pk_1` and `pk_2`. 

For the testing versions, we made the following changes: 
* Slight modification to the recursive design, to omit the usage of the period
* Introduction of a macro based construction, to allow for an fixed size array based (instead of vector based) solution

### CODE REVIEW

File: kes-mmm-sumed25519/Cargo.toml [hunk 1]
```diff
@@ -7,16 +7,20 @@ categories = [ "cryptography" ]
...
 keywords = [ "Crypto", "KES", "Ed25519", "MMM", "Sum" ]
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-
 [dependencies]
-ed25519-dalek = "1.0.0-pre.3"
-rand_core = "0.5"
-sha2 = "0.8"
+ed25519-dalek = "1"
+rand = { version = "0.7" }
```
Comment (vincenthz) [inline kes-mmm-sumed25519/Cargo.toml]: why ? this just add 6 dependencies (e.g. `log`) for absolutely nothing.

Comment (iquerejeta) [reply kes-mmm-sumed25519/Cargo.toml]: left over, used it for a few more things that ended up deleting. I'll go back to rand_core 👍

File: kes-mmm-sumed25519/src/lib.rs [hunk 2]
```diff
@@ -12,12 +13,22 @@ extern crate quickcheck;
...
 
 mod common;
+#[macro_use]
+mod single_kes;
+pub mod erasable_buffer;
+pub mod errors;
 pub mod sumed25519;
+mod traits;
 
+/// Module to handle version environment variables.
```
Comment (vincenthz) [inline kes-mmm-sumed25519/src/lib.rs:16]: it's not about returning the environment variables, it's about returning the *versions* of the library that happens to be in the environment at the time of the compilation

**Comment by iquerejeta** [comment]
@vincenthz I've basically changed the whole repo (omitting for now the C and Haskell interface). If you could have a view, that'd be great. Otherwise I'll run this through the internal team of auditing. 

In summary, what we achieve with these changes is to be compatible with both supported versions in main-net. 

1. Pre-babbage, `KesSig`, is a naive KES signature, where both public keys are stored in the signature.
2. Post-babbage, `CompactSig` where the optimization of using a single key in the sig was introduced. This efficiency trick was basically taken from the previous implementation here 👍 

Tests under `tests` ensure that this implementation is compatible with what we currently use in the node, by generating some test vectors from the [cardano-base](https://github.com/input-output-hk/cardano-base) library.

**Comment by vincenthz** [comment]
at this rate, I think it would make more sense to just create a new repository specially as the rewrite is providing a cardano solution (e.g. a `cardano-kes` repository), and keeping this repository as historical artefact with a implementation of a cryptographic paper.

**Comment by iquerejeta** [comment]
Agreed. Closing this PR in favour of a new repo dedicated for the [Cardano compatible version](https://github.com/input-output-hk/kes).

**Pull request closed without merging**